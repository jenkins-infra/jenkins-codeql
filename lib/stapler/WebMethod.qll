import java

import jenkins.ProductionCallable

class WebMethod extends Method, ProductionCallable {
    WebMethod() {
        // TODO make this nicer and more comprehensive, e.g. all parameters need to have one of several annotations or be one of several types

        // Stapler basic requirements for a method to be dispatchable
        // and include Jenkins DoActionFilter since 2.138.3
        this.isPublic()
        and
        not(this.hasAnnotation("jenkins.security.stapler", "StaplerNotDispatchable"))
        and
        (
            // Ensure the web method is properly annotated
            this.hasAnnotation("jenkins.security.stapler", "StaplerDispatchable")
            or
            this.hasAnnotation("org.kohsuke.stapler", "WebMethod")
            or
            this.getAnAnnotation().getType().hasAnnotation("org.kohsuke.stapler.interceptor", "InterceptorAnnotation")
            or
            this.getName().indexOf("do") = 0
        )
        and not exists(Parameter p | this.getAParameter() = p
            // Ensure no parameter exists without expected type or annotation
            and
            (
                not p.getType().hasName("StaplerRequest")
                and not p.getType().hasName("StaplerResponse")
                and not p.getType().hasName("HttpServletRequest")
                and not p.getType().hasName("HttpServletResponse")
                and not p.getType().hasName("ServletRequest")
                and not p.getType().hasName("ServletResponse")
                and not p.getType().hasName("RequestImpl")
                and not p.getType().hasName("ResponseImpl")
            )
            and not exists(Annotation a | p.getAnAnnotation() = a
                and
                (
                    a.getType().hasName("QueryParameter")
                    or
                    a.getType().hasName("AncestorInPath")
                    or
                    a.getType().hasName("Header")
                    or
                    a.getType().hasName("JsonBody")
                    or
                    a.getType().hasName("SubmittedForm")
                )
            )
        )
    }
}
