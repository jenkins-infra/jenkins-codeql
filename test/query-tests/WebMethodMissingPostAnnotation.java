import org.kohsuke.stapler.AncestorInPath;
import org.kohsuke.stapler.HttpResponse;
import org.kohsuke.stapler.QueryParameter;

import hudson.security.AccessControlled;
import hudson.security.Permission;
import hudson.util.ListBoxModel;

import java.net.URL;

public class WebMethodMissingPostAnnotation {
    public void doCheckWhatever(@AncestorInPath AccessControlled ac, @QueryParameter String value) throws Exception {
        ac.checkPermission(Permission.CREATE);
        new URL(value).openConnection();
    }
    public ListBoxModel doFillWhatever() {
        ListBoxModel box = new ListBoxModel();
        box.add("whatever");
        return box;
    }
}
