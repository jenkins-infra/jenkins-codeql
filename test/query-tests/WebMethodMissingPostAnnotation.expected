| FileFromRestOfPath.java:7:17:7:31 | doPathTraversal | Potential CSRF vulnerability: If FileFromRestOfPath#doPathTraversal connects to user-specified URLs, modifies state, or is expensive to run, it should be annotated with @POST or @RequirePOST |
| FileFromRestOfPath.java:12:19:12:27 | doNotCare | Potential CSRF vulnerability: If FileFromRestOfPath#doNotCare connects to user-specified URLs, modifies state, or is expensive to run, it should be annotated with @POST or @RequirePOST |
| WebMethodMissingPostAnnotation.java:12:17:12:31 | doCheckWhatever | Potential CSRF vulnerability: If WebMethodMissingPostAnnotation#doCheckWhatever connects to user-specified URLs, modifies state, or is expensive to run, it should be annotated with @POST or @RequirePOST |
